Sources/Workspace/Workspace.swift

```diff
@@ -495,90 +540,106 @@ public class Workspace {
-        let currentManifests = loadDependencyManifests(rootManifests: rootManifests, diagnostics: diagnostics)
-        guard !diagnostics.hasErrors() else { return }
-
-        // Compute constraints with the new pin and try to resolve
-        // dependencies. We only commit the pin if the dependencies can be
-        // resolved with new constraints.
-        //
-        // The constraints consist of three things:
-        // * Unversioned constraints for edited packages.
-        // * Root manifest contraints without pins.
-        // * Exisiting pins except the dependency we're currently pinning.
-        // * The constraint for the new pin we're trying to add.
-        var constraints = currentManifests.unversionedConstraints()
-        constraints += rootManifests.flatMap({ $0.package.dependencyConstraints() })
-        constraints += pinsStore.createConstraints().filter({ $0.identifier != dependency.repository })
-        constraints.append(
-            RepositoryPackageConstraint(
-                container: dependency.repository, requirement: requirement))
-
-        // Resolve the dependencies.
-        let results = resolveDependencies(constraints: constraints, diagnostics: diagnostics)
-        guard !diagnostics.hasErrors() else { return }
-
-        // Update the checkouts based on new dependency resolution.
-        try updateCheckouts(with: results)
-
-        // Get the updated dependency.
-        let managedDependencies = try self.managedDependencies.load()
-        let newDependency = managedDependencies[dependency.repository]!
-
-        // Assert that the dependency is at the pinned checkout state now.
-        if case .checkout(let checkoutState) = newDependency.state {
-            assert(checkoutState.requirement() == requirement)
-        } else {
-            assertionFailure()
+    ) {
+        do {
+            guard case .checkout(let currentState) = dependency.state else {
+                throw WorkspaceError.cantPinDependencyInEditMode(dependency: dependency)
+            }
+            
+            // Compute the requirement.
+            let requirement: RepositoryPackageConstraint.Requirement
+            if let version = version {
+                requirement = .versionSet(.exact(version))
+            } else if let branch = branch {
+                requirement = .revision(branch)
+            } else if let revision = revision {
+                requirement = .revision(revision)
+            } else {
+                requirement = currentState.requirement()
+            }
+            
+            let pinsStore = try self.pinsStore.load()
+            let rootManifests = loadRootManifests(packages: rootPackages, diagnostics: diagnostics)
+            let currentManifests = loadDependencyManifests(rootManifests: rootManifests, diagnostics: diagnostics)
+            guard !diagnostics.hasErrors() else { return }
```

Comment:
Instead of mixing do..catch and hasError() checks, we should use emit and return unless then entire method uses do..try. It is confusing to use both variants in same method, and I find the extra nesting annoying :/